mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -856,9 +699,87 @@ int64_t mlir::rock::getMaxVectorization(ArrayAttr transforms, uint32_t dim,
   return result;
 }
 
-AffineMap mlir::rock::composeTransforms(ArrayAttr transforms) {
+/// Embed operations can create some scenarios that lead to the need to
+/// check if their output falls with the expected range. The first is any
+/// subtraction. The second is the case where
+/// [coifficient] * [max size of upper dim] > [lower bound] for any upper
+/// dimension.
+static bool embedCanBeInvalid(TransformMapAttr map, TransformAttr op) {
+  assert(op.getType() == TransformType::Embed);
+  int64_t lowerBound = map.getLowerBounds()[op.getLowerDims()[0]];
+  ArrayRef<int64_t> dimSizes = map.getUpperBounds();
+  return llvm::any_of(llvm::zip(op.getParams(), op.getUpperDims()),
+                      [&](const auto &pair) -> bool {
+                        int64_t coefficient = std::get<0>(pair);
+                        uint32_t dim = std::get<1>(pair);
+                        return (coefficient < 0) ||
+                               ((dimSizes[dim] * coefficient) > lowerBound);
+                      });
+}
+
+bool mlir::rock::mapImpactsValidity(TransformMapAttr map) {
+  bool result = false;
+  for (TransformAttr op : map.getOps()) {
+    TransformType type = op.getType();
+    ArrayRef<int64_t> params = op.getParams();
+    if (type == TransformType::Pad) {
+      for (size_t i = 0, e = params.size(); i < e; i += 2) {
+        // Trivial padding doesn't impact validity
+        result |= (params[i] != 0 || params[i + 1] != 0);
+      }
+    } else if (type == TransformType::Embed) {
+      result |= embedCanBeInvalid(map, op);
+    }
+  }
+  return result;
+}
+
+Value mlir::rock::updateValidityAfter(OpBuilder &b, Location loc,
```

Comment:
I think docs for these functions would be useful

Reply:
They're in the header